src/libpmemobj/tx.c

```diff
@@ -1116,6 +1178,99 @@ tx_realloc_common(struct tx *tx, PMEMoid oid, size_t size, uint64_t type_num,
 }
 
 /*
+ * pmemobj_tx_ctx_init -- initializes the transaction context
+ */
+void
+pmemobj_tx_ctx_init(void)
+{
+	LOG(3, NULL);
+	int result = pthread_once(&tx_ctx_once, tx_ctx_create_key);
+	if (result) {
+		errno = result;
+		FATAL("!pthread_once");
+	}
+
+	/*
+	 * Workaround Helgrind's bug:
+	 * https://bugs.kde.org/show_bug.cgi?id=337735
+	 */
+	VALGRIND_ANNOTATE_HAPPENS_AFTER(&tx_ctx_once);
+}
+
+/*
+ * pmemobj_tx_ctx_fini -- destroys the transaction context
+ */
+void
+pmemobj_tx_ctx_fini(void)
+{
+	LOG(3, NULL);
+	void *ctx = pthread_getspecific(tx_ctx_key);
+	if (ctx != NULL) {
+		Free(ctx);
+		(void) pthread_setspecific(tx_ctx_key, NULL);
+	}
+}
+
+/*
+ * pmemobj_tx_setctx -- sets new transaction context
+ *
+ * Upon success the old transaction context is stored under old_ctx.
+ * A pointer passed as old_ctx cannot be NULL.
+ */
+int
+pmemobj_tx_ctx_set(struct pobj_tx_ctx *new_ctx, struct pobj_tx_ctx **old_ctx)
+{
+	LOG(3, "new_ctx %p, old_ctx %p", new_ctx, old_ctx);
+	if (old_ctx == NULL) {
+		ERR("pointer for old transaction context cannot be NULL");
+		return -1;
+	}
+
+	struct tx *old_tx = get_tx_ctx();
+	if (old_tx != NULL && old_tx->pop != NULL)
+		old_tx->lane_nest_count = lane_detach(old_tx->pop);
+
+	struct tx *new_tx = (struct tx *)new_ctx;
+	if (new_tx != NULL && new_tx->pop == NULL) {
+		ERR("NULL pop");
+		return -1;
+	}
+
+	if (set_tx_ctx(new_tx))
+		return -1;
+
+	*old_ctx = (struct pobj_tx_ctx *)old_tx;
+
+	if (new_tx != NULL)
+		lane_attach(new_tx->pop, new_tx->lane, new_tx->lane_nest_count);
+
+	return 0;
+}
+
+/*
+ * pmemobj_tx_restorectx -- restores transaction context
+ *
+ * The current transaction context is replaced and then freed
+ */
+int
+pmemobj_tx_ctx_restore(struct pobj_tx_ctx *ctx)
+{
+	LOG(3, "ctx %p", ctx);
+	if (ctx == NULL) {
+		ERR("transaction context cannot be NULL");
+		return -1;
+	}
+
+	struct pobj_tx_ctx *old_ctx;
+	if (pmemobj_tx_ctx_set(ctx, &old_ctx)) {
+		ERR("!pmemobj_tx_setctx");
+		return -1;
+	}
+	Free(old_ctx);
```

Comment:
Why?
I don't understand this function.

The "pobj_tx_ctx" should be created in pmemobj_tx_begin() and freed in pmemobj_tx_end(), always and unconditionally.